File: Review-07-Java-Toys/Example-3/src/main/java/edu/odu/cs/cs350/examples/Roster.java
package edu.odu.cs.cs350.examples;

import java.util.Set;
import java.util.LinkedHashSet;

/**
 * A class roster listing all students enrolled
 * in a course
 */
public class Roster implements Cloneable {
    public static final int DEFAULT_MAX_STUDENTS = 10;

    private String courseNum;
    private int    enrollLimit;

    private LinkedHashSet<Student> students;

    /**
     * Create a class roster with a limit
     * of DEFAULT_MAX_STUDENTS `Student`s
     * and the course number set to
     * "CS 150"
     */
    public Roster()
    {
        this.courseNum   = "CS 150";
        this.enrollLimit = DEFAULT_MAX_STUDENTS;

        students = new LinkedHashSet<Student>();
    }

    /**
     * Create a class roster with a specified 
     * enrollment limit and course number
     *
     * @param l enrollment limit
     * @param c course number 
     */
    public Roster(int l, String c)
    {
        this.courseNum   = c;
        this.enrollLimit = l;

        students = new LinkedHashSet<Student>();
    }

    /**
     * Retrieve the course number
     */
    public String getCourseNum()
    {
        return this.courseNum;
    }

    /**
     * Change the course number
     *
     * @param n desired course number
     */
    public void setCourseNum(String n)
    {
        this.courseNum = n;
    }

    /**
     * Retrieve the enrollment limit
     */
    public int getEnrollLimit()
    {
        return this.enrollLimit;
    }

    /**
     * Change the enrollment limit
     *
     * @param n desired limit
     */
    public void setEnrollLimit(int n)
    {
        this.enrollLimit = n;
    }

    /**
     * Attempt to enroll a Student in the course
     *
     * Rules: 
     *   1- A student can not be added if the enrollment
     *     limit has been reached.
     *   2- A can not be added to a roster multiple times 
     *
     * @param stu Student to enroll
     *
     * @return true if the Student was successfully enrolled
     *    in the course (added to the roster)
     */
    public boolean enroll(Student stu)
    {
        // limit has been reached
        if(students.size() == enrollLimit) {
            return false;
        }

        // Student was previously added to the roster
        if(students.contains(stu)) {
            return false;
        }

        students.add(stu);
        return true;
    }

    /**
     * Retrieve the number of enrolled students
     */
    public int numEnrolled()
    {
        return students.size();
    }

    /**
     * Return a collection of students in
     * the order they were enrolled (added)
     *
     * @return Set of enrolled students. If no students
     *     have been added to the roster, the set will be 
     *     empty.
     */
    public Set<Student> listEnrolledStudents()
    {
        return this.students;
    }

    /**
     * Compare 2 `Student`s based on name
     *
     * @param rhs the other (right-hand-side) student object
     */
    @Override
    public boolean equals(Object rhs)
    {
        if(!(rhs instanceof Roster)) {
            return false;
        }

        Roster rhsRoster = (Roster) rhs;

        if (!this.courseNum.equals(rhsRoster.courseNum)) {
            return false;
        }

        if (this.enrollLimit != rhsRoster.enrollLimit) {
            return false;
        }

        if (!this.students.equals(rhsRoster.students)) {
            return false;
        }        

        return true;
    }

    /**
     * Return a hashcode
     */
    @Override
    public int hashCode()
    {
        int hc = courseNum.hashCode();

        hc += enrollLimit;
        hc += students.hashCode();

        return hc;
    }

    /**
     * Return a (deep) copy of this object.
     */
    @Override
    public Object clone()
    {
        Roster cpy = new Roster(this.enrollLimit, this.courseNum);

        // Now add the students
        for (Student stu : this.students) {
            cpy.students.add(stu);
        }

        //return new Student(name);
        return cpy;
    }

    /**
     * Generate a String containing the course number,
     * number of enrolled students, enrollment limit, and the names
     * of all enrolled students.
     */
    @Override
    public String toString()
    {
        StringBuilder bld = new StringBuilder();

        bld.append(this.courseNum);
        bld.append(
            String.format(" -> %2d of %2d (%4.2f%% full)\n",
                          students.size(),
                          enrollLimit,
                          100.0 * students.size() / enrollLimit)
        );

        for (Student s : students) {
            bld.append("  -" + s + "\n");
        }

        return bld.toString();
    }
}

File: Review-00-Foundational-Concepts/Classes-Composition-Python/examples/roster.py
from __future__ import annotations

import copy

from examples.student import Student

DEFAULT_MAX_STUDENTS: int = 10


class Roster:
    """
    A class roster listing all students enrolled in a course.
    """

    def __init__(self, l: int = DEFAULT_MAX_STUDENTS, c: str = "CS 150"):
        """
        Create a class roster with a specified
        enrollment limit and course number.

        Args:
            l enrollment limit
            c course number
        """
        self.course_num = c
        self.enroll_limit = l

        self.students = set()

    @property
    def course_number(self):
        return self.course_num

    @course_number.setter
    def course_number(self, n: int):
        self.course_num = n

    @property
    def get_enrollment_limit(self):
        return self.enroll_limit

    def set_enroll_limit(self, n: int):
        """
        Change the enrollment limit.
         *
        @param n desired limit
        """

        self.enroll_limit = n

    def enroll(self, stu: Student) -> bool:
        """
        Attempt to enroll a Student in the course.

        Rules:
          1- A student can not be added if the enrollment
            limit has been reached.
          2- A can not be added to a roster multiple times
         *

        Args:
          stu Student to enroll

        Returns:
           true if the Student was successfully enrolled
           in the course (added to the roster)
        """
        # limit has been reached
        if len(self.students) == self.enroll_limit:
            return False

        # Student was previously added to the roster
        if stu in self.students:
            return False

        self.students.add(stu)
        return True

    def num_enrolled(self):
        """
        Retrieve the number of enrolled students.
        """

        return len(self.students)

    def __len__(self):
        """
        Retrieve the number of enrolled students.
        """
        return len(self.students)

    def list_enrolled_students(self):
        """
        Return a collection of students in
        the order they were enrolled (added).
         *
        @return Set of enrolled students. If no students
            have been added to the roster, the set will be
            empty.
        """

        return self.students

    def __eq__(self, rhs) -> bool:
        """
        Compare 2 `Student`s based on name.
         *
        @param rhs the other (right-hand-side) student object
        """
        if not isinstance(rhs, Roster):
            return False

        if self.course_num != rhs.course_num:
            return False

        if self.enroll_limit != rhs.enroll_limit:
            return False

        if self.students != rhs.students:
            return False

        return True

    def __hash__(self):
        hc = hash(self.course_num)

        hc += self.enroll_limit

        # Python sets can not be hashed (3.11.2)
        #  hc += hash(self.students)
        hc += sum(hash(stu) for stu in self.students)

        return hc

    def __deepcopy__(self, memo):
        cpy = Roster(self.enroll_limit, self.course_num)

        # Now add the students
        for stu in self.students:
            cpy.students.add(stu)

        return cpy

    def __str__(self):
        """
        Generate a String containing the course number, number of enrolled
        students, enrollment limit, and the names of all enrolled students.
        """

        percent_full = 100.0 * len(self.students) / self.enroll_limit

        return (
            self.course_num
            + f" -> {len(self.students):>2} of {self.enroll_limit:>2}"
            + f" ({percent_full}% full)\n"
            + "\n".join((f"  - {stu}" for stu in self.students))
        )
